enhancements/node/node-operator.md

```diff
@@ -0,0 +1,316 @@
+---
+title: node-operator
+authors:
+  - "@saschagrunert"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2021-06-22
+last-updated: 2021-06-22
+status: provisional
+see-also: {}
+replaces: {}
+superseded-by: {}
+---
+
+# Node Operator (NOP)
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The overall goal of creating this new operator in OpenShift is the encapsulation
+of CRI-O and kubelet related configurations, metrics and alerts within a single
+domain and namespace.This provides on one hand a higher flexibility for the Node
+team, while on the other hand the dependency on external teams gets reduced. In
+the long-term the Node team would only rely on OpenShift internal APIs to manage
+the container runtime and kubelet related configurations. Additional deployments
+for proxying runtime metrics or providing telemetry access are part of this
+operator, too.
+
+## Motivation
+
+Most of the node related configurations are today part of the [Machine Config
+Operator (MCO)][mco]. Because the MCO got bigger over time, their long-term goal
+is to provide an API for machine related configuration purposes. The new
+operator would use this API in a later migration step.
+
+The CRI-O metrics are today served by using a [`kubelet`
+ServiceMonitor][kubelet-monitor] and a corresponding Prometheus relabeling. This
+happens as part of the [Cluster Monitoring Operator (CMO)][cmo] project. The CMO
+team also aims to provide an API for external teams to integrate their
+metrics retrieval into the OpenShift cluster monitoring. In the same way as for
+the MCO, the new node operator would use this API to integrate the runtime
+metrics retrieval natively into OpenShift by using a dedicated ServiceMonitor
+secured by TLS and RBAC.
+
+[mco]: https://github.com/openshift/machine-config-operator
+[cmo]: https://github.com/openshift/cluster-monitoring-operator
+[kubelet-monitor]: https://github.com/openshift/cluster-monitoring-operator/blob/8ec92e1/assets/control-plane/service-monitor-kubelet.yaml#L95-L116
+
+### Goals
+
+- Near-term:
+  - Create a new Node owned project to be prepared for future changes of the MCO
+    and CMO.
+  - Build in CRI-O metrics retrieval by using TLS and RBAC with the help of a
+    kube-rbac-proxy DaemonSet.
+  - Include TLS certificate rotation (on disk as well as in a secret) for the
+    CRI-O metrics.
+- Long-term:
+  - Migrate the kubelet and CRI-O configurations into the operator.
```

Comment:
Aside from metrics configuration and export below I don't see any other examples of kubelet and cri-o configuration mentioned. Is that because it's just not in scope for the initial proposal?

Once we get there I'm curious how we'd address coupling of configuration with the binaries presently managed by MCO updating CoreOS. Not sure if that's in scope or not.

Reply:
^ that is a question I have, as well as how are we going to pull the NO images? right now we use podman to pull the mcd and run a one-off to configure before cri-o starts. would we use cri-o to pull the NO images (thus requiring a cri-o restart), or would we do a similar thing with podman, then configure cri-o

Reply:
> Aside from metrics configuration and export below I don't see any other examples of kubelet and cri-o configuration mentioned. Is that because it's just not in scope for the initial proposal?

Yes, the idea was that the Node Operator only holds the configuration, while we use an API of the MCO later to apply the config. 

> ^ that is a question I have, as well as how are we going to pull the NO images?

I would say that we use the standard way and let CRI-O pull the image. How we apply the configuration later on would be still managed by the MCO (API), right?